File: rocketmq-client/src/main/java/org/apache/rocketmq/client/impl/consumer/ConsumeTask.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.client.impl.consumer;

import com.google.common.base.Stopwatch;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;
import org.apache.rocketmq.client.consumer.ConsumeContext;
import org.apache.rocketmq.client.consumer.ConsumeStatus;
import org.apache.rocketmq.client.consumer.listener.MessageListener;
import org.apache.rocketmq.client.message.MessageExt;
import org.apache.rocketmq.client.message.MessageHookPoint;
import org.apache.rocketmq.client.message.MessageHookPointStatus;
import org.apache.rocketmq.client.message.MessageInterceptor;
import org.apache.rocketmq.client.message.MessageInterceptorContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ConsumeTask implements Callable<ConsumeStatus> {
    private static final Logger log = LoggerFactory.getLogger(ConsumeTask.class);

    private final MessageInterceptor interceptor;
    private final MessageListener messageListener;
    private final List<MessageExt> messageExtList;

    public ConsumeTask(MessageInterceptor interceptor, MessageListener messageListener,
                       List<MessageExt> messageExtList) {
        this.interceptor = interceptor;
        this.messageListener = messageListener;
        this.messageExtList = messageExtList;
    }

    @Override
    public ConsumeStatus call() {
        // intercept before message consumption.
        for (MessageExt messageExt : messageExtList) {
            final MessageInterceptorContext context =
                    MessageInterceptorContext.builder().setAttempt(messageExt.getDeliveryAttempt()).build();
            interceptor.intercept(MessageHookPoint.PRE_MESSAGE_CONSUMPTION, messageExt, context);
        }

        ConsumeStatus status;

        final ConsumeContext consumeContext = new ConsumeContext();
        final Stopwatch stopwatch = Stopwatch.createStarted();
        try {
            status = messageListener.consume(messageExtList, consumeContext);
        } catch (Throwable t) {
            status = ConsumeStatus.ERROR;
            log.error("Message listener raised an exception while consuming messages.", t);
        }
        if (null == status) {
            log.error("Message listener returns NPE for consume status");
            status = ConsumeStatus.ERROR;
        }

        // intercept after message consumption.
        final TimeUnit timeUnit = MessageInterceptor.DEFAULT_TIME_UNIT;
        final long duration = stopwatch.elapsed(timeUnit);
        final MessageHookPointStatus pointStatus = ConsumeStatus.OK.equals(status) ? MessageHookPointStatus.OK :
                                                   MessageHookPointStatus.ERROR;
        final int batchSize = messageExtList.size();
        for (MessageExt messageExt : messageExtList) {
            final int attempt = messageExt.getDeliveryAttempt();
            final MessageInterceptorContext context = MessageInterceptorContext.builder()
                                                                               .setAttempt(attempt)
                                                                               .setDuration(duration)
                                                                               .setTimeUnit(timeUnit)
                                                                               .setBatchSize(batchSize)
                                                                               .setStatus(pointStatus).build();
            interceptor.intercept(MessageHookPoint.POST_MESSAGE_CONSUMPTION, messageExt, context);
        }
        return status;
    }
}

File: nodejs/src/consumer/ConsumeTask.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MessageView } from '../message';
import {
  Attribute,
  AttributeKey,
  MessageHookPoints,
  MessageHookPointsStatus,
  MessageInterceptor,
  MessageInterceptorContextImpl,
} from '../hook';
import { ConsumeResult } from './ConsumeResult';
import { MessageListener } from './MessageListener';

// Context keys exposed to interceptors, mirroring the Java ConsumeTask keys.
export const REMOTE_ADDR_CONTEXT_KEY = AttributeKey.create<string>('remote_address');
export const MESSAGE_VIEW_CONTEXT_KEY = AttributeKey.create<MessageView>('message_view');
export const CONSUMER_GROUP_CONTEXT_KEY = AttributeKey.create<string>('consumer_group');
export const CONSUME_ERROR_CONTEXT_KEY = AttributeKey.create<unknown>('consume_error');

export class ConsumeTask {
  readonly #messageListener: MessageListener;
  readonly #messageView: MessageView;
  readonly #consumerGroup?: string;
  readonly #messageInterceptor?: MessageInterceptor;

  constructor(_clientId: string, messageListener: MessageListener, messageView: MessageView,
    messageInterceptor?: MessageInterceptor, consumerGroup?: string) {
    this.#messageListener = messageListener;
    this.#messageView = messageView;
    this.#messageInterceptor = messageInterceptor;
    this.#consumerGroup = consumerGroup;
  }

  async call(): Promise<ConsumeResult> {
    const generalMessages = [ this.#messageView ];
    let context = new MessageInterceptorContextImpl(MessageHookPoints.CONSUME);
    context.putAttribute(MESSAGE_VIEW_CONTEXT_KEY, Attribute.create(this.#messageView));
    if (this.#consumerGroup) {
      context.putAttribute(CONSUMER_GROUP_CONTEXT_KEY, Attribute.create(this.#consumerGroup));
    }
    let throwable: unknown;
    let consumeResult: ConsumeResult;
    if (this.#messageInterceptor) {
      this.#messageInterceptor.doBefore(context, generalMessages);
    }
    try {
      consumeResult = await this.#messageListener.consume(this.#messageView);
    } catch (e) {
      // Message listener raised an exception while consuming messages
      throwable = e;
      consumeResult = ConsumeResult.FAILURE;
    }
    const status = consumeResult === ConsumeResult.SUCCESS ? MessageHookPointsStatus.OK : MessageHookPointsStatus.ERROR;
    context = MessageInterceptorContextImpl.withStatus(context, status);
    if (throwable) {
      context.putAttribute(CONSUME_ERROR_CONTEXT_KEY, Attribute.create(throwable));
    }
    if (this.#messageInterceptor) {
      this.#messageInterceptor.doAfter(context, generalMessages);
    }
    return consumeResult;
  }
}
